### PR TITLE
Sd 965 inverse firmware search rest

### PR DESCRIPTION
**Pull request by jstucke**
- added `inverse` parameter to `/rest/firmware` for inverse parent firmware search
- refactoring

### CODE REVIEW

File: src/test/unit/helperFunctions/test_rest.py [hunk 1]
```diff
@@ -64,40 +67,46 @@ def test_convert_rest_request_succeeds(data):
...
+def test_get_boolean_from_request(get_request_parameter_function, name):
+    assert not get_request_parameter_function(None)
 
     with pytest.raises(ValueError):
-        get_summary_flag(dict(summary='bad_string'))
+        get_request_parameter_function({name: 'bad_string'})
 
     with pytest.raises(ValueError):
-        get_summary_flag(dict(summary='2'))
+        get_request_parameter_function({name: '2'})
```
Comment (dorpvom) [inline src/test/unit/helperFunctions/test_rest.py:101]: Redundant? (I know it's not coming from you)

File: src/web_interface/rest/rest_firmware.py [hunk 2]
```diff
@@ -80,26 +80,31 @@ def _update_analysis(self, uid, update):
...
 
-            return success_message(dict(uids=uids), self.URL, dict(offset=offset, limit=limit, query=query, recursive=recursive))
-        except Exception:
-            return error_message('Unknown exception on request', self.URL, dict(offset=offset, limit=limit, query=query, recursive=recursive))
+    @staticmethod
+    def _get_parameters_from_request(request_parameters):
+        query = get_query(request_parameters)
+        recursive = get_recursive_flag(request_parameters)
+        inverse = get_inverse_flag(request_parameters)
+        paging = get_paging(request.args)
```
Comment (dorpvom) [inline src/web_interface/rest/rest_firmware.py:102]: `offset, limit = ` would make this more readable in line **84**

File: src/web_interface/rest/rest_firmware.py [hunk 3]
```diff
@@ -80,26 +80,31 @@ def _update_analysis(self, uid, update):
...
+            request_data = {k: request.args.get(k) for k in ['query', 'limit', 'offset', 'recursive', 'inverse']}
+            return error_message(str(value_error), self.URL, request_data=request_data)
 
+        parameters = dict(offset=offset, limit=limit, query=query, recursive=recursive, inverse=inverse)
         try:
             with ConnectTo(FrontEndDbInterface, self.config) as connection:
-                uids = connection.rest_get_firmware_uids(offset=offset, limit=limit, query=query, recursive=recursive)
+                uids = connection.rest_get_firmware_uids(**parameters)
+            return success_message(dict(uids=uids), self.URL, parameters)
+        except Exception:  # pylint: disable=broad-except
```
Comment (dorpvom) [inline src/web_interface/rest/rest_firmware.py:94]: See above. Maybe we can replace the broad except

File: src/web_interface/rest/rest_firmware.py [hunk 4]
```diff
@@ -80,26 +80,31 @@ def _update_analysis(self, uid, update):
...
-        except Exception:
-            return error_message('Unknown exception on request', self.URL, dict(offset=offset, limit=limit, query=query, recursive=recursive))
+    @staticmethod
+    def _get_parameters_from_request(request_parameters):
+        query = get_query(request_parameters)
+        recursive = get_recursive_flag(request_parameters)
+        inverse = get_inverse_flag(request_parameters)
+        paging = get_paging(request.args)
+        if recursive and not query:
+            raise ValueError('recursive search is only permissible with non-empty query')
```
Comment (dorpvom) [inline src/web_interface/rest/rest_firmware.py:104]: I'd choose a capital first letter on these error messages

File: src/storage/db_interface_frontend.py [hunk 5]
```diff
@@ -241,9 +241,9 @@ def create_analysis_structure(self):
 
         return visualize_complete_tree(stripped_field_strings)
 
-    def rest_get_firmware_uids(self, offset, limit, query=None, recursive=False):
+    def rest_get_firmware_uids(self, offset, limit, query=None, recursive=False, inverse=False):
         if recursive:
-            return self.generic_search(search_dict=query, skip=offset, limit=limit, only_fo_parent_firmware=True)
+            return self.generic_search(search_dict=query, skip=offset, limit=limit, only_fo_parent_firmware=True, inverted=inverse)
```
[the remark this answers is not in the document]
Comment (jstucke) [reply src/storage/db_interface_frontend.py:246]: changed to inverted everywhere (I hope)

File: src/web_interface/rest/rest_file_object.py [hunk 6]
```diff
@@ -27,23 +27,20 @@ def _fit_file_object(file_object):
...
         try:
             with ConnectTo(FrontEndDbInterface, self.config) as connection:
-                uids = connection.rest_get_file_object_uids(offset=offset, limit=limit, query=query)
-
-            return success_message(dict(uids=uids), self.URL, dict(offset=offset, limit=limit, query=query))
-        except Exception:
-            return error_message('Unknown exception on request', self.URL, dict(offset=offset, limit=limit, query=query))
+                uids = connection.rest_get_file_object_uids(**parameters)
+            return success_message(dict(uids=uids), self.URL, parameters)
+        except Exception:  # pylint: disable=broad-except
```
[the remark this answers is not in the document]
Comment (jstucke) [reply src/web_interface/rest/rest_file_object.py:42]: changed to `PyMongoError`

File: src/test/unit/helperFunctions/test_rest.py (hunk 1 above)
Comment (jstucke) [reply src/test/unit/helperFunctions/test_rest.py:101]: I thougt so too but kept it anyway -> removed

File: src/web_interface/rest/rest_firmware.py (hunk 2 above)
Comment (jstucke) [reply src/web_interface/rest/rest_firmware.py:102]: changed

File: src/web_interface/rest/rest_firmware.py (hunk 3 above)
Comment (jstucke) [reply src/web_interface/rest/rest_firmware.py:94]: changed to `PyMongoError`

File: src/web_interface/rest/rest_firmware.py (hunk 4 above)
Comment (jstucke) [reply src/web_interface/rest/rest_firmware.py:104]: changed

**Pull request merged**